src/libraries/Microsoft.PowerFx.LanguageServerProtocol/JsonRpc/LanguageServerInput.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System.Text.Json;
+
+namespace Microsoft.PowerFx.LanguageServerProtocol
+{
+    /// <summary>
+    /// Represents the input to the language server.
+    /// </summary>
+    public class LanguageServerInput
+    {
+        /// <summary>
+        /// Id of the request. Applicable only for request messages.
+        /// </summary>
+        public string Id { get; private set; }
+
+        /// <summary>
+        /// Method of the request or notification.
+        /// </summary>
+        public string Method { get; private set; }
+
+        /// <summary>
+        /// The jsonrpc version.
+        /// </summary>
+        public string Jsonrpc { get; private set; }
+
+        /// <summary>
+        /// Parameters of the request or notification.
+        /// </summary>
+        public string Params { get; private set; }
+        
+        /// <summary>
+        /// Parses the input json payload to <see cref="LanguageServerInput"/>.
+        /// </summary>
+        /// <param name="jsonRpcPayload">Raw Payload.</param>
+        /// <returns>Parsed Input.</returns>
+        public static LanguageServerInput Parse(string jsonRpcPayload)
+        {
+            var input = new LanguageServerInput();
+            try
+            {
+                using (var document = JsonDocument.Parse(jsonRpcPayload))
+                {
+                    if (document == null)
+                    {
+                        return input;
+                    }
+
+                    var root = document.RootElement;
+                    if (root.TryGetProperty("id", out var idElement))
+                    {
+                        input.Id = idElement.GetString();
+                    }
+
+                    if (root.TryGetProperty("method", out var methodElement))
+                    {
+                        input.Method = methodElement.GetString();
+                    }
+
+                    if (root.TryGetProperty("jsonrpc", out var jsonrpcElement))
+                    {
+                        input.Jsonrpc = jsonrpcElement.GetString();
+                    }
+
+                    if (root.TryGetProperty("params", out var paramsElement))
+                    {
+                        input.Params = paramsElement.GetRawText();
+                    }

```

Comment:
Could we use strong parsing here?
Deserialize<LanguageServerInput>(jsonRpcPAyload)

Reply:
Params differ in structure depending on what lsp method we are trying to handle. The parsing of params is left up to the handlers as they know how to parse their respective params structure. If it intellisense, params is deserialized to completion params, if it is signature help, it is deserialized to signature params and so on. This was the case in old sdk as welll. If we want to strongly parse it here then we would need to switch case on lsp method and select the right type. 

It makes sense for each handlers to take the responsibility of the params. Doing it here seems infeasible and violation of separation of concerncs

Reply:
Let's chat more and I can explain.